Supports global selection for all devices

chunhtai opened the pull request:
Support for mouse drag selection and mobile selection

https://github.com/flutter/flutter/issues/81839

design doc: http://flutter.dev/go/global-selection

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+/// of the collected selectables.
+///
+/// If the [enabled] is false, this container does not collect selectables in
+/// the subtree, it will also hide them from other [SelectionContainer]
```
The paragraph above was saying that they are never visible even if enabled is true. This makes it sound like they are only hidden if enabled is false?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+/// [onSelectablesChange] callback and is responsible for the selection update
+/// of the collected selectables.
+///
+/// If the [enabled] is false, this container does not collect selectables in
```
nit: drop the "the" before enabled.

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+class SelectionContainer extends StatefulWidget {
+  /// creates a selection container to collect the [Selectable]s in the subtree.
+  ///
+  /// If the [enabled] is true, the [onSelectablesChange] must not be null.
```
drop the "the"

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// Called when the selectables this widget collected has changed.
```
What constitutes a change? Is that just a selectable getting added or removed? Something else?

goderbauer commented on packages/flutter/lib/src/services/selection.dart:
```diff
@@ -0,0 +1,218 @@
+    return 'DragSelectionEndEvent()';
+  }
+}
+
```
nit: remove extra blank line

goderbauer commented on packages/flutter/lib/src/services/selection.dart:
```diff
@@ -0,0 +1,218 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Why are these defined in the services library?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  /// Use this method to extract the collected selectables in the subtree.
+  final ValueChanged<List<Selectable>>? onSelectablesChange;
+
+  /// Gets the immediate ancestor [SelectionRegistrar] of the [BuildContext].
```
The doc comment of this class should probably introduce SelectionRegistrar as a concept.

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+class _SelectionContainerState extends State<SelectionContainer> implements SelectionRegistrar {
+  /// Registered selectable in screen order.
+  List<Selectable> get _selectables {
+    if(_isCachedSelectablesDirty) {
```
nit: space after if

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+    return _cachedSelectables;
+  }
+  bool _isCachedSelectablesDirty = true;
+  List<Selectable> _cachedSelectables = const <Selectable>[];
```
Maybe a little clearer: cachedSortedSelectables

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+
+  /// Called when the selectables this widget collected has changed.
+  ///
+  /// Use this method to extract the collected selectables in the subtree.
```
document that the selectables are given in screen order in the list?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+
+  DragSelectionStartEvent? _currentDragStart;
+
+  /// Updates the selectables this updater manages.
```
Document that it expects the selectables in screen order (I think?)

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  /// Gets the list selectables this updater is responsible for updating.
+  ///
+  /// The subclasses is responsible for updating the selection when the
+  /// selectables in the list has changed.
```
document expected order?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  }
+
+  /// Copies the selected contents of all the selectables.
+  Object? copy() {
```
Should this be typed `String?`? Seems to be the only possible return type?

chunhtai replied:
It may not be the case if we support multimedia, but i guess we can change that later?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+
+  /// Initialize the selection of the selectable children.
+  ///
+  /// The goal is to find which selectable child the selection starts on
```
missing `.`

goderbauer reviewed (commented):
Haven't reviewed it all, just some comments from an initial read-through.

goderbauer commented:
This looks pretty interesting. It's nice to see that we can do this without adding things to RenderObject.

What wasn't clear to me from my initial read-through is why we need `SelectionResult.next`/`SelectionResult.previous`. Why can that not just be determined based on the rect of a Selectable?

chunhtai commented on packages/flutter/lib/src/services/selection.dart:
```diff
@@ -0,0 +1,218 @@
+
+import 'dart:ui';
+
+import 'package:flutter/cupertino.dart';
```
I forgot to clean this up, this was added automatically by the ide...

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+///
+/// This class optimize the selection update by keeping track the current
+/// selectable that contains the selection edge.
+abstract class MultiSelectableSelectionUpdaterBase {
```
This is also used by scrollable.dart, I can probably keep StaticMultiSelectableSelectionUpdater private though.

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  }
+
+  Map<ShortcutActivator, Intent> _getPlatformShortcuts() {
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.iOS:
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
```
I need to double check with iOS, but other use control+a/v/c to copy and paste. Only mac use cmd+a/v/c

goderbauer replied:
Maybe remove the windows from the name of the constant then.

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  }
+
+  void _cancelSelection() {
+    _endDragSelection();
```
will fix it, we should make it not allowed

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+    SingleActivator(LogicalKeyboardKey.keyA, meta: true) : _SelectAllIntent(),
+  };
+
+  static const Map<ShortcutActivator, Intent> _kWindowsShortcuts = <ShortcutActivator, Intent>{
```
Oh, sorry, one is control, the other is meta. Never mind.

chunhtai replied:
mac use meta:true which is cmd, the other uses control: true

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+
+/// A container that collects the [Selectable]s in the subtree.
+///
+/// This widget creates its own [SelectionContainer] to collect the
```
ah it was a typo, I made SelectionRegistrarScope private and rename all the doc reference to SelectionContainer. Didn't expect this to happen hahaha.

goderbauer commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -298,6 +301,148 @@ class RenderParagraph extends RenderBox
+    late final SelectionResult result;
+    if (event is DragSelectionEvent) {
+      if (event is DragSelectionStartEvent) {
+        _dragSelectionStartInGlobalCoordinate = event.offset;
```
Wouldn't we want to hold on to the start point in local coordinates und use that for the calculations in _updateDragSelection? 

I am thinking about a scenario where you start the selection with a pointer down in a large body of text and while the pointer is down you scroll a surrounding scrollable. I think that scroll needs to extend the selection and the selection needs to start at the later that the pointer touched when it cam down first. Due to the scroll, that letter will have moved in the global coordinate system, though.

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+/// selectables in the subtree. These selectables are not visible to any other
+/// [SelectionContainer] above this widget. The creator of this widget
+/// can extract the selectables this container collected using the
+/// [onSelectablesChange] callback and is responsible for the selection update
```
will added some doc example, the sole purpose of using SelectionContainer with enabled=true is that you want to manage the selectables subtree by yourself, it doesn't make sense to use the  SelectionContainer without providing the callback.

I have thought about different alternative

1. use globalkey to grab selectables from SelectionContainerState, but it may be hard to detect when the selectables has changed.
2. Use the Inherited way to get the selectable from SelectionContainer, but that means the widget that actually updates the selectables needs to be below the SelectionContainer, it makes code very confusing.

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,586 @@
+  /// should start before all of its children. Returns [SelectionResult.next]
+  /// if the selection should start after all of its children.
+  SelectionResult _initSelection(DragSelectionUpdateEvent event) {
+    for (int index = 0; index < selectables.length; index += 1){
```
nit: space before {

chunhtai commented:
> This looks pretty interesting. It's nice to see that we can do this without adding things to RenderObject.
> 
> What wasn't clear to me from my initial read-through is why we need `SelectionResult.next`/`SelectionResult.previous`. Why can that not just be determined based on the rect of a Selectable?

The biggest concern is select to scroll. Imagine a container with 3 child [A, B, C]. If B is scrollable and the drag selection goes to C,  IMO we can't send update event to C until B finishes scrolling. We can also makes B returns some kind of flag to tell the container not to proceed and get rid of SelectionResult, but I think the current implementation is more systematic since there maybe more cases where selectable itself have opinion on the selection order.

The other concern is keyboard selection. Imagine developer press shit+ down, how does the container know which selectable should be the new selection edge. With SelectionResult, the selectable itself can tell the container whether the shift down should reach the next selectable or not.

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/95226

skia-gold commented:
Gold has detected about 2 new digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/95226

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 11.
View them at https://flutter-gold.skia.org/cl/github/95226

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 12.
View them at https://flutter-gold.skia.org/cl/github/95226

chunhtai committed:
Support global selection

chunhtai commented:
please ignore the failing test for now, It is because I want to hide the selection specific logic in RenderParagraph by making a private subclass of it, and it failed the describe node test. I am thinking whether there is way around it or i will have to expose the API

chunhtai commented:
I will add more tests. This PR now supports both mouse selection and mobile selection. I figure i may need to tear up bunch of API in Selectable to separate both them into two PR, which may make it hard to comprehend.  So I keep them in the same PR.

chunhtai commented:
also cc @jonahwilliams

jonahwilliams commented on packages/flutter/lib/src/material/selection.dart:
```diff
@@ -0,0 +1,927 @@
+    _selectable?.dispatchSelectionEvent(const SelectAllSelectionEvent());
+  }
+
+  Future<void> _copy([Intent? intent]) async {
```
How do we eventually expand support towards copy/paste for rich data types? i.e. images and things

chunhtai replied:
we can have a within app clipboard, but I am not sure how we can store them into system clipboard as they usually seems to store different data type with different API. for example windows have Clipboard.setText and Clipboard.setImage.

jonahwilliams replied:
I would think about how we could expand this in the future to include some metadata that would indicate the type of data to be copied. I don't necessarily think you need to worry about it initially, but it would be cool to support

chunhtai replied:
yes, I think that is a good idea. will think about how we can do that

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,962 @@
+/// selectable under a [SelectionArea].
+///
+/// The [child] must be a [RichText].
+class TypographySelectionAdapter extends SingleChildRenderObjectWidget {
```
@justinmc I am curious how you feel about this approach

To refactor the EditableText, the renderEditable needs to implement `RenderSelectableTypography` and EditableText needs to wrap the widget tree with this adapter.

These API can apply to non text renderobject too as long as they implement the `RenderSelectableTypography`, although a lot of class it uses has the keyword `Text*` in it, e.g TextSelection, TextRange, TextPosition, ideally these should have a more general names.

chunhtai replied:
`RenderSelectableTypography` is defined in rendering/selection.dart

chunhtai replied:
I am not entirely sure how to correctly handle RenderParagraph with WidgetSpan. The easiest we can do is to either select the entire widgetspan as one character, or make it possible to select partial content in the WidgetSpan. I know our internal customers like to inline text within text using WidgetSpan, and the former behavior may not suffice their needs.

justinmc replied:
That seems super straightforward and clean to do, but I have some questions.

Is the ultimate point of having RenderEditable implement RenderSelectableTypography to allow the reuse of selection code?  I'll look more closely at what that would look like if so.

Also what's the status of TextFields mixing with SelectionAreas now?  I left repro steps in my main review about a bug related to this, but maybe it's not fully supported yet or something.

chunhtai replied:
If we want to support partial selection in WidgetSpan, this interface will not be useful because the Text widget will need to be some kind of SelectionContainer or the RenderParagraph needs to delegate the selection event down to individual widgetspan, but I think we don't need to support that in first version?

>Is the ultimate point of having RenderEditable implement RenderSelectableTypography to allow the reuse of selection code? I'll look more closely at what that would look like if so.

That is the current plan yes, but I started to worry about rich text editing, are we going to support something like widgetspan? if so, are we going to treat them as one character? can they be partially selected or modify? If so this interface won't work.

> Also what's the status of TextFields mixing with SelectionAreas now? I left repro steps in my main review about a bug related to this, but maybe it's not fully supported yet or something.

Right now they are two separate systems, so you can't select across text field with other text item. I think we should keep them separate even after refactoring. The selection can't go in or out from the textfield to other widget, otherwise text editing may be confusing. WDYT?

chunhtai replied:
Otherwise, IMO this API is perfect for renderobject without child and want to be selectable, it is far easier to implement this than the Selectable abstract class.

justinmc commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -106,11 +107,16 @@ class RenderParagraph extends RenderBox
 
+  static final Paint _selectionPaint = Paint()
+    ..style = PaintingStyle.fill
+    ..color = const Color(0xAF6694e8);
```
The color used by EditableText is more complicated than this, it's based on platform (e.g. [here](https://github.com/flutter/flutter/blob/e5f9d5bd26982922dca64acc9066e32fe806e12a/packages/flutter/lib/src/material/text_field.dart#L1170)).  Maybe we could pull that code out of TextField etc. and reuse it everywhere.

chunhtai replied:
ah right, we should add a theme for this

justinmc reviewed (commented):
I know you're still working on parts of this but I thought I'd leave a review.  Feel free to ignore any of this that is not finalized yet.

Question that came up: Do we have a plan for the future of SelectableText?  Should we deprecate it in favor of Text in a SelectionArea?

One potential bug I found: A TextButton inside of a SelectionArea will have its text selectable directly using the text cursor (the mouse cursor didn't show up in the screenshot, but it was the text cursor and not the pointer mouse cursor).  I would have expected that it would be selectable all at once by dragging before the button to after the button, but not individually inside the button like this.  I wonder if there are any other widgets that use Text that would cause a problem like this.

<img width="87" alt="Screen Shot 2022-01-28 at 3 12 07 PM" src="https://user-images.githubusercontent.com/389558/151634300-c8c15a34-d78e-4204-8100-793cf4eabab0.png">

I imagine long term, when Flutter supports more than plain text in the clipboard, we'll want to have widgets like Image implement this so that they can show a selection highlight.  Is that right?  Any other widgets that should show selection highlight?

Another corner case I found is pressing cmd+a to select all within a text field.  It selects only outside of the text field and not the text field text too.  (Or maybe mixing with EditableText isn't supported yet?)

https://user-images.githubusercontent.com/389558/151626674-725befc1-0db8-45e6-818f-783166b3e4dd.mov

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Flutter Demo'),
        ),
        body: Center(
          child: Padding(
            padding: const EdgeInsets.all(32.0),
            child: SelectionArea(
              child: Column(
                children: const <Widget>[
                  Text('I am some text.'),
                  TextField(
                    maxLines: 3,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                    ),
                  ),
                  Text('I even more text.'),
                  Text('Last one.'),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

chunhtai commented:
> Question that came up: Do we have a plan for the future of SelectableText? Should we deprecate it in favor of Text in a SelectionArea?

I think we don't need to deprecate it, we can make it build a Text widget and wrap it with its own SelectionArea if there isn't one above the widget

jonahwilliams commented:
Just a drive by response to the above: I don't think selectable text should exist. It was an efficient shortcut to leverage the existing edit text capabilties to give folks _something_, but Ideally we would deprecate and remove it.

Hixie commented on packages/flutter/test/widgets/selection_test.dart:
```diff
@@ -0,0 +1,7 @@
+// found in the LICENSE file.
+
+void main() {
+  // TODO(chunhtai): add tests for selection container.
```
this should happen before we land this, or we shouldn't land the file. no point landing the file with just the todo :-)

chunhtai replied:
yes, I am in the progress of adding more tests, please ignore the test file for now.  I would like to get some feedback on code structure first in the meanwhile in case I may need to rewrite a bunch of tests if API changes as the result of addressing feedbacks.

chunhtai commented:
> Just a drive by response to the above: I don't think selectable text should exist. It was an efficient shortcut to leverage the existing edit text capabilties to give folks _something_, but Ideally we would deprecate and remove it.

Yes, that sounds good. I take my previous response back

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+
+  @override
+  _RenderSelectableContainerWidget updateRenderObject(BuildContext context, _RenderSelectableContainerWidget renderObject) {
+    assert(renderObject.delegate == delegate && renderObject.containerHasSelectableContent == hasSelectableContent);
```
What guarantees that these don't change?

chunhtai replied:
good catch

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+  }
+
+  @override
+  _RenderSelectableContainerWidget updateRenderObject(BuildContext context, _RenderSelectableContainerWidget renderObject) {
```
Why change the return type from void?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+
+/// An abstract base class for updating multiple selectable children.
+///
+/// This class optimize the selection update by keeping track the [Selectable]s
```
keeping track _of_ the

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+///
+/// This class optimize the selection update by keeping track the [Selectable]s
+/// that currently contain the selection edges.
+abstract class MultiSelectableDelegateBase with ChangeNotifier implements SelectionContainerDelegate {
```
document what the change notifier notifies about.

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+import 'layer.dart';
+import 'object.dart';
+
+/// The result after handling a selection event. This is used by the parent
```
nit: first paragraph should just be one sentence (just add two '\n' after the ".").

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+import 'object.dart';
+
+/// The result after handling a selection event. This is used by the parent
+/// of multiple [Selectable]s to determine how a selection expand across its
```
expand -> expands

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+/// selectable children.
+enum SelectionResult {
+  /// There is nothing left to select forward, and further selection should
+  /// extend to the next selectable in screen order.
```
everywhere: change selectable to [Selectable]?

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// There is nothing left to select forward, and further selection should
+  /// extend to the next selectable in screen order.
+  next,
+  /// Selection does not reach this selectable and should look previous
```
should look _at the_ previous

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  previous,
+  /// Selection ends in this selectable.
+  ///
+  /// part of the selectable may or may not be selected, but there are still
```
there are -> there _is_

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// part of the selectable may or may not be selected, but there are still
+  /// content to select forward or backward.
+  end,
+  /// The result can't be determine in this frame.
```
determined

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// selection handles.
+  ///
+  /// This handler is responsible for painting the leader layers with the
+  /// given layer links if they are not null. It is possible only one layer is
```
It is possible _that_ only

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// non-null if this handler is only responsible for painting that non-null
+  /// layer link.
+  ///
+  /// If you are implementing this method as a result of mixing [Selectable] to
```
see style guide about "you".

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// See also:
+  ///  * [HandleLayersPainterMixin]: which paint layer links automatically for
+  ///    the mixer.
+  void pushHandleLayers(LayerLink? startHandle, LayerLink? endHandle);
```
Where should these be placed?

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// Return `null` if nothing is selected.
+  Object? getSelectedContent();
+
+  /// handles the [SelectionEvent] sent to this object.
```
handles -> Handles

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+/// the [SelectionRegisteree] to subscribe to the [SelectionRegistrar]
+/// automatically.
+///
+/// The [Selectable] also paints [LayerLink]s of selection handles in a
```
My understanding is that the Selectable doesn't do this and instead whoever mixes this class in has to implement that feature?

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+/// [hasSelectableContent]. This mixin only registers itself with the
+/// [registrar] if there is selectable content in this [Selectable].
+mixin SelectionRegisteree on Selectable {
+
```
nit: remove blank line

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+  /// Whether there is selectable content in the selectable.
+  ///
+  /// Modify the value of this [ValueNotifier] to notify this mixin about
+  /// content changes.
```
How do you do that? Do you just switch it between false and true whenever the content changes? Seems ... odd.

goderbauer replied:
Or do you only change this when it changes from having content to not having content?

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+        ? SelectionResult.next
+        : SelectionResult.previous;
+  }
+
```
nit: remove one blank line

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+
+  @override
+  String toString() {
+    return 'SelectionStartEdgeUpdateEvent(offset: $globalPosition)';
```
use `objectRuntimeType` here and below.

goderbauer commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,562 @@
+/// The geometry of current selection.
+///
+/// This includes detail, such as the location of the selection start or end,
+/// line height...etc. This information is used for drawing selection handles
```
line hight, etc.

goderbauer commented on packages/flutter/test/material/selection_test.dart:
```diff
@@ -0,0 +1,103 @@
+void main() {
+  group('SelectionArea', () {
+    testWidgets('SectionArea mouse selection sends correct event', (WidgetTester tester) async {
+      // await tester.pumpWidget(
```
commented out code

goderbauer commented on packages/flutter/test/material/selection_test.dart:
```diff
@@ -0,0 +1,103 @@
+  });
+}
+
+// class SelectionSpy extends SingleChildRenderObjectWidget {
```
commented out code

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+///
+/// To disable a selection for a subtree, wraps the subtree using
+/// [SelectionContainer.disabled].
+class SelectionContainer extends StatefulWidget {
```
> Is this the main/central class of text selection?

Or is SelectionArea the main class?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+  /// The owner of this delegate.
+  ///
+  /// The delegate queries layout geometry through the [owner].
+  final State<StatefulWidget> owner;
```
This is odd. Any State object can be the owner?

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+    required Widget child,
+  }) : super(key: key, child: child);
+
+  /// The registrar.
```
this comment doesn't add any information over the property name

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+  /// The registrar.
+  final SelectionRegistrar registrar;
+
+  /// whether this enabled
```
same.

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+  }
+}
+
+/// The adapter to handles the selections from the [SelectionArea] for
```
nit: grammar of this sentence. (actually not sure what you want to say)

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+/// By Wrapping a [RichText] with this widget, the [RichText] will be
+/// selectable under a [SelectionArea].
+///
+/// The [child] must be a [RichText].
```
If it must be a RichText, why is it typed as Widget? But also first bullet of: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-interleaving-multiple-concepts-together

goderbauer commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+/// The adapter to handles the selections from the [SelectionArea] for
+/// [RichText].
+///
+/// By Wrapping a [RichText] with this widget, the [RichText] will be
```
What's the motivation for having this?

goderbauer reviewed (commented):
Since this is a huuuge PR I haven't fully gotten to all the pieces yet and only scanned over some parts to get an overview.

goderbauer commented:
Some questions that should be answered somewhere in the doc:
* For global selection, what pieces does an app developer need to care about?
* What do I need to do to enable/disable global selection for my app? What do I do to enable it on some platforms (e.g. web) but not others (mobile, some desktop platforms, I assume).
* Some kind of overview of how all these pieces fit together.

chunhtai commented on packages/flutter/lib/src/widgets/selection.dart:
```diff
@@ -0,0 +1,977 @@
+    if (scope?.enabled != true) {
+      return null;
+    }
+    return scope?.registrar;
```
it can still be null because the previous `if` check for inequality

chunhtai replied:
well I guess i can do the other way around and get rid of the `!`

chunhtai commented on packages/flutter/lib/src/services/selection.dart:
```diff
@@ -0,0 +1,218 @@
+    if (point.dy > targetRect.bottom)
+      return SelectionResult.next;
+    return point.dx >= targetRect.right
+        ? SelectionResult.next
```
the selection result next/previous is in screen order, the LTR and RTL does not matter in this case.

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/95226

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -816,6 +834,466 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+
+  @override
+  void didChangeSelectables() {
+    // print('didchange selectable ${selectables.map((dynamic e) => e.paragraph.text.toPlainText())}');
```
Here's a print statement to remove.

justinmc reviewed (commented):
This is still a partial review, but I wanted to post it now and follow up later.  I want to continue to play around with it and make sure I understand it.  I left some questions below to help me.

Also I tried this out and tried to break it, and I wanted to list a few potential QA issues:

  1. It's possible to display two selections at once by selecting something in a SelectionArea, and then pressing tab to move to a TextField:

<img width="513" alt="Screen Shot 2022-04-20 at 4 45 32 PM" src="https://user-images.githubusercontent.com/389558/164342473-f1640e6f-ee1d-44de-a8b9-36d6b9c13332.png">

Should we strictly only allow one selection ever?  Maybe there is a more fundamental way to ensure that.  Maybe in the long term outside of this PR, though...

  2. Unintuitive selection order?  I think I would expect it to select first to the bottom of the first column, then top to bottom of the second column. Here is a codepen of how it looks on the web: https://codepen.io/justinmc/pen/bGaZwXV


https://user-images.githubusercontent.com/389558/164343474-24575b54-b499-42ad-8d75-8da18160ff66.mov


This is basically the following widgets:

```dart
Row(
  children: [
    SizedBox(
      width: 200.0,
      child: Column(
        children: [ Text('Hello col1 lorem ipsum etc....'), Text('Hello col1'), Text('Hello col1'), ],
      ),
    ),
    SizedBox(
      width: 200.0,
      child: Column(
        children: [ Text('Hello col2.'), Text('Hello col2'), Text('Hello col2'), ],
      ),
    ),
  ],
),

chunhtai commented:
> Unintuitive selection order? I think I would expect it to select first to the bottom of the first column, then top to bottom of the second column. Here is a codepen of how it looks on the web: https://codepen.io/justinmc/pen/bGaZwXV

I have not enforce selection order on column and row yet, right now everything is in screen order. This can be fixed by wrapping every child in the Row widget with a SelectionContainer, so that it enforce every selectable in one child is selected before moving on to the next. but i plan to do that in a later pr.

chunhtai commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,883 @@
+/// If a [Selectable] mixer contains a dispose method, i.e [RenderObject] or
+/// [State], it would conflict with [ChangeNotifier]. Use this mixin makes sure
+/// the dispose method is called correctly.
+mixin SelectableChangeNotifier on Selectable {
```
I think there is way around it, let me try

chunhtai commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,886 @@
+/// A mixin to paint handle layers automatically based on selection geometry.
+///
+/// To use this mixin, subclass must call super when overriding [paint] method.
+mixin HandleLayersPainterMixin on Selectable, RenderBox {
```
ah it was not used after a refactor, we may need this at some point once we convert more widget to be selectable, remove it for now

chunhtai commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,886 @@
+  /// [SelectionResult.next]. This is useful when the drag offset is outside of
+  /// the target rectangle or the target does not contain any selectable
+  /// contents; therefore, the selection can't end in this [Selectable].
+  static SelectionResult selectionBasedOnRect(Rect targetRect, Offset point) {
```
it is not, it is the rect of a single [selectable]

chunhtai replied:
oh yes, I thought you mean the widget SelectionArea.

chunhtai commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,886 @@
+/// An abstract base class for selection events.
+///
+/// This should not be directly used. To handle a selection event, it should
+/// be downward cast to a specific subclass.
```
the correct term should be downcast https://en.wikipedia.org/wiki/Downcasting#:~:text=In%20class%2Dbased%20programming%2C%20downcasting,one%20of%20its%20derived%20classes.
will update

chunhtai commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,886 @@
+/// [SelectionArea]'s gesture recognizer for mouse devices, or the end selection
+/// handle has been dragged to a new location. The [globalPosition] contains the
+/// new location of the selection end edge.
+class SelectionEndEdgeUpdateEvent extends SelectionEdgeUpdateEvent {
```
If I remember correctly the intent/actions is to bubble up to be handle by the parents. This is the other way around

chunhtai commented on packages/flutter/lib/src/widgets/selection_area.dart:
```diff
@@ -0,0 +1,1344 @@
+  /// Create a new [SelectionArea] widget.
+  ///
+  /// The [selectionControls] is used for building selection handles and toolbar
+  /// for mobile devices. If it is null, the platform specific selection
```
do you know why are they private? I can create a MaterialSelectionArea now, or I can make these class public. What do you think makes most sense

chunhtai commented:
> It's possible to display two selections at once by selecting something in a SelectionArea, and then pressing tab to move to a TextField:

yes I think that is something we should do, and it is probably can be fixed once textfield is converted to use global selection?

I can fix it right now, but it will further complex the API, my plan is to handle it once i refactor textfield. WDYT?

justinmc commented on packages/flutter/lib/src/widgets/text.dart:
```diff
@@ -342,10 +344,37 @@ class DefaultTextHeightBehavior extends InheritedTheme {
 ///
+/// ## Selections
+///
+/// [Text] is selectable by default if it is under a [SelectionArea]. To disable
```
Does this PR add a SelectionArea anywhere by default?  Like in Scaffold or MaterialApp or something?

goderbauer replied:
Let's add this information to the doc here (i.e. that you have to manually wrap your tree in this to enable it)

justinmc commented on packages/flutter/lib/src/widgets/text.dart:
```diff
@@ -342,10 +344,37 @@ class DefaultTextHeightBehavior extends InheritedTheme {
+/// ## Selections
+///
+/// [Text] is selectable by default if it is under a [SelectionArea]. To disable
+/// the selection, wraps the Text or part of the subtree with
```
"wraps" => "wrap"

justinmc commented on packages/flutter/lib/src/widgets/text.dart:
````diff
@@ -342,10 +344,37 @@ class DefaultTextHeightBehavior extends InheritedTheme {
+///
+/// ![](https://flutter.github.io/assets-for-api-docs/assets/widgets/rich_text.png)
+///
+/// ```dart
````
Good useful example :+1:

justinmc commented on packages/flutter/lib/src/widgets/text.dart:
```diff
@@ -372,6 +401,7 @@ class Text extends StatelessWidget {
     this.semanticsLabel,
     this.textWidthBasis,
     this.textHeightBehavior,
+    this.selectionColor = const Color(0xAF6694e8),
```
Should this be a constant somewhere?  I wonder where this color came from.  Here and below.

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -29,38 +29,6 @@ export 'package:flutter/services.dart' show TextSelectionDelegate;
-///
-///  * [TextDirection], which discusses left-to-right and right-to-left text in
-///    more detail.
-enum TextSelectionHandleType {
```
Heads up that this could be a breaking change if someone is using this and not importing the rendering library.

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+    TestDefaultBinaryMessengerBinding.instance!.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.platform, null);
+  });
+
+  testWidgets('mouse can select multiple widget', (WidgetTester tester) async {
```
"widget" => "widgets"

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+    expect(paragraph50.selections[0], const TextSelection(baseOffset: 2, extentOffset: 4));
+  });
+
+  testWidgets('can select all non-mac', (WidgetTester tester) async {
```
Nit: Maybe saying "Apple" instead of "Mac" is more accurate since it includes iOS.

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+    ));
+    await tester.pumpAndSettle();
+
+  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS, TargetPlatform.macOS }));
```
Is this test incomplete or am I misunderstanding?

chunhtai replied:
oh it looks like i missed it, nice catch!

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+    expect(paragraph97.selections[0], const TextSelection(baseOffset: 0, extentOffset: 7));
+    expect(paragraph96.selections[0], const TextSelection(baseOffset: 0, extentOffset: 7));
+    await gesture.up();
+  });
```
Do the selection handles show up in reaction to touch events even on desktop platforms?

I think this might be how EditableText works now too, no need to change it if so.

chunhtai replied:
yes if you creates EditableText with materialTextSelectionControls, it will show up for desktop too

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+  });
+
+  group('Complex cases', () {
+    testWidgets('selection start outside of the scrollable', (WidgetTester tester) async {
```
"start" => "starts"

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+      expect(innerParagraph24.selections[0], const TextSelection(baseOffset: 0, extentOffset: 2));
+    });
+
+    testWidgets('can copy off screen selection - mac', (WidgetTester tester) async {
```
Good idea to test this, I hadn't thought of it.

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,641 @@
+
+      final Map<String, dynamic> clipboardData = mockClipboard.clipboardData as Map<String, dynamic>;
+      expect(clipboardData['text'], 'em 0It');
+    }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.android, TargetPlatform.windows, TargetPlatform.linux, TargetPlatform.fuchsia }));
```
Nit: Is the only difference between this and the "mac" test above the keyboard keys?  You could also do a `switch` on the defaultTargetPlatform and do the correct keystrokes there in one test.

justinmc replied:
Same for other tests elsewhere.

chunhtai replied:
I prefer to do it this way, using the switch of defaultTargetPlatform may create test that can fail on certain platform only, it may be hard to debug locally if you don't have the device

chunhtai replied:
I will keep it as is, let me know if you think otherwise

justinmc replied:
I'm on board with keeping it as-is :+1:

justinmc commented on packages/flutter/test/widgets/selection_area_test.dart:
```diff
@@ -0,0 +1,1021 @@
+      await gesture.up();
+    });
+
+    testWidgets('mouse can select multiple widget', (WidgetTester tester) async {
```
"widget" => "widgets"

justinmc commented on packages/flutter/test/widgets/selection_area_test.dart:
```diff
@@ -0,0 +1,1021 @@
+      },
+      variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.android, TargetPlatform.windows, TargetPlatform.linux, TargetPlatform.fuchsia }),
+      skip: isBrowser, // https://github.com/flutter/flutter/issues/61020
+    );
```
What if the WidgetSpan didn't contain a Text?  Would it be represented as a placeholder character or left out?  Maybe add a test for that if it's useful and there's not already one...

chunhtai replied:
added

justinmc commented on packages/flutter/test/widgets/selection_container_test.dart:
```diff
@@ -0,0 +1,112 @@
+
+void main() {
+  group('SelectionContainer', () {
+    testWidgets('works', (WidgetTester tester) async {
```
Maybe "updates its registrar and delegate based on the number of selectables" or something like that.

justinmc reviewed (commented):
Thanks for waiting for my review!  Most comments are grammar nits, but there's an architectural question and some other things in there.

I asked for examples in a few comments, it might be worth it to add a free-standing example in the examples/ directory if you agree.

> > It's possible to display two selections at once by selecting something in a SelectionArea, and then pressing tab to move to a TextField:

> yes I think that is something we should do, and it is probably can be fixed once textfield is converted to use global selection?

> I can fix it right now, but it will further complex the API, my plan is to handle it once i refactor textfield. WDYT?

Got it.  No need to fully refactor how selection is maintained in this PR.  But if there is an easy fix for that double selection bug in the short term then we should do it.  It wasn't too hard to discover that bug so I imagine users will too.

> do you know why are they private? I can create a MaterialSelectionArea now, or I can make these class public. What do you think makes most sense

([comment](https://github.com/flutter/flutter/pull/95226#discussion_r856550018))  They were purposely made private when I created the desktop right click menus because they weren't yet finalized, so we didn't want users depending on them.  I think they still shouldn't be made public yet, not until I've integrated them with @gspencergoog's nested menus.

Some more QA observations:

  1. The positioning of the menu seems off here:

![Screenshot from 2022-04-27 16-25-01](https://user-images.githubusercontent.com/389558/165647241-85b10a79-31d0-4840-abb7-fbd892957434.png)

  2. Multiple SelectionAreas allow multiple simultaneous selections.  Is that desirable?

chunhtai committed:
addressing comments

chunhtai reviewed (commented):
I addressed most of the comments.

What is still missing:
1. adding an example for selection area
2. adding an example for selection container
3. figure out what to do with private desktop selectionControls.

chunhtai commented:
> I addressed most of the comments.
> 
> What is still missing:
> 
> 1. adding an example for selection area
> 2. adding an example for selection container
> 3. figure out what to do with private desktop selectionControls.

3, It looks like people can still use https://api.flutter.dev/flutter/material/desktopTextSelectionControls.html, so i think there should be no problem?

chunhtai commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,200 @@
+/// [SelectionEvent]s received by this container are sent to the [delegate] using
+/// the [SelectionHandler] API of the delegate.
+///
+/// To use this widget, set the [registrar] this container received the
```
added example

chunhtai commented:
Added example and fix the multiple selection area cases.

>The positioning of the menu seems off here:
![image](https://user-images.githubusercontent.com/47866232/167211409-09fcd4e9-5495-4e6b-941b-f1cc1bf3cf32.png)

This is due to the calculation logic in the buildToolbar, whenever there is multi line selection the toolbar will be put into the middle, I was thinking this may be fixable after the context menu refactor.

justinmc commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -146,6 +152,107 @@ class RenderParagraph extends RenderBox
+  }
+
+  List<_SelectableFragment> _getSelectableFragments() {
+    final String plainText = text.toPlainText(includeSemanticsLabels: false);
```
If a user copies do they get placeholder chars in their clipboard?

chunhtai replied:
They won't, in line 226 the TextRange is [start, end).

Added a new test `widget span can copy correctly` in `selectable_region_test.dart` to verify that

justinmc commented on packages/flutter/lib/src/widgets/selection_area.dart:
```diff
@@ -0,0 +1,1349 @@
+    // 2. the selectable returns next when looking backward.
+    while (newIndex < selectables.length &&
+        newIndex >= 0 && finalResult == null) {
+      currentSelectableResult = dispatchSelectionEventToChild(selectables[newIndex], event);
```
Thanks for the clear explanation, I'm on board for both reasons :+1:.  I really hadn't thought about selection being asynchronous, wow.

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,545 @@
+  /// The selection handle type that should be used at the selection point.
+  ///
+  /// This is used for building the mobile selection handle.
+  final TextSelectionHandleType handleType;
```
Agreed, good catch.

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,545 @@
+  }
+}
+
+/// Which type of selection handle to be displayed.
```
Ah sorry, bonus docs cleanup :)

chunhtai committed:
add new test

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+/// This object is responsible for drawing selection highlight.
+///
+/// In order to receive the selection event, the mixer need to register
+/// themselves to [SelectionRegistrar]s. Use the
```
Nit: I think it's more clear to say "itself" instead of "themselves" to be clear that we're referring to a single mixer.

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+///
+/// This object is responsible for drawing selection highlight.
+///
+/// In order to receive the selection event, the mixer need to register
```
"need" => "needs"

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+/// mix the [SelectionRegistrant] to subscribe to the [SelectionRegistrar]
+/// automatically.
+///
+/// The mixer also need to paints [LayerLink]s of selection handles in a
```
"need" => "needs"

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+  /// The size of this [Selectable].
+  Size get size;
+
+  /// The dispose method to enable mixer to use [SelectionRegistrant].
```
"mixer" => "the mixer"

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+    if (value == _registrar)
+      return;
+    if (value == null) {
+      // when registrar go from non-null to null;
```
Capitalize "When" here and below, and maybe periods instead of semicolons if that wasn't on purpose.

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,570 @@
+
+/// The type of selection handle to be displayed.
+///
+/// With mixed-direction text, both handles may be the same type. Examples:
```
These examples here are good, thanks for explaining the edge case!

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -802,6 +820,465 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+  /// Creates a auto scroller that scrolls the [scrollable].
+  EdgeDraggingAutoScroller(this.scrollable, {this.onScrollViewScrolled, this.velocityScalar = _kDefaultAutoScrollVelocityScalar});
+
+  // An eyeball value for a smooth scrolling experience.
```
"eyeball" => "eyeballed"

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -802,6 +820,465 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+/// This updater handles the case where the selectables change frequently, and
+/// it optimizes toward scrolling updates.
+///
+/// This updater keeps track of the drag start offset related to scroll origin
```
Nit: "This updater" => "It"

justinmc reviewed (commented):
I think this is almost ready to go, but I left one comment about the Shortcuts thing.  I want to make sure it's easy for the user to override, or at least as easy as it is practical for us.

I should also give a more detailed review on some of the docs next week, I skipped over reading them all in the latter part of my review.

QA:

I noticed that if I copy everything from your [selection_area.dart example](https://github.com/flutter/flutter/pull/95226/files#diff-a90e85a748afbb4ab24927fbf2eb51e19a41f906d330a1f06ead0704233cafae), then when I paste it, it doesn't include newlines between the row items.  I feel like it would if I did this on a regular non-Flutter website.  Not sure how important that is, though.

chunhtai commented on examples/api/lib/material/selection_area/custom_selectable.dart:
```diff
@@ -0,0 +1,243 @@
+  }
+}
+
+class MySelectableAdapter extends StatelessWidget {
```
This widget is very specific. I can think of a more advanced mixin to let customer use that fill out most of the boilerplate. We can do that once we got more use cases in the future.

chunhtai commented on packages/flutter/lib/src/material/selection_area.dart:
```diff
@@ -0,0 +1,272 @@
+  @override
+  Widget build(BuildContext context) {
+    TextSelectionControls? controls = widget.selectionControls;
+    late Map<ShortcutActivator, Intent> shortcuts;
```
Although this may be a rare case, people may be changing platform in Them data?

chunhtai replied:
will be removed after https://github.com/flutter/flutter/pull/103377 merges

chunhtai commented on packages/flutter/lib/src/material/selection_area.dart:
```diff
@@ -0,0 +1,272 @@
+        behavior: HitTestBehavior.translucent,
+        excludeFromSemantics: true,
+        child: Shortcuts(
+          shortcuts: shortcuts,
```
I don't think there is an easy way unless SelectionArea receives the shortcuts as input. Originally I thought I can use DefaultTextEditiingShortcuts, but realize the web shortcut disabled select all and copy.

goderbauer commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,207 @@
+    super.initState();
+    widget.delegate._selectionContainerContext = context;
+    widget.delegate.addListener(_onSelectionGeometryChange);
+    _notifier = ValueNotifier<SelectionGeometry>(widget.delegate.value);
```
The delegate implements SelectionHandler, which itself implements ValueListenable. Why do we need to wrap this in another ValueNotifier? Can you just add/remove listeners to the delegate directly and access its value?

chunhtai replied:
right, not sure what I was thinking.

goderbauer commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,207 @@
+  /// This is useful when retrieving layout geometry of the container.
+  @protected
+  BuildContext get selectionContainerContext => _selectionContainerContext!;
+  BuildContext? _selectionContainerContext;
```
The way the BuildContext is injected into this Delegate seems ... iffy. What is that actually needed for?

chunhtai replied:
it needs the layout information

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+  @override
+  void initState() {
+    super.initState();
+    _selectionDelegate = _SelectableRegionContainerDelegate();
```
nit: could just be instantiated in line 177 and then the late could be replaced with final?

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+  /// If there is already an existing select end edge, this method will update
+  /// it to the new location.
+  ///
+  /// If the `continuous` is set to true and the update causes scrolling, the
```
remove the first "the"

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+  /// it to the new location.
+  ///
+  /// If the `continuous` is set to true and the update causes scrolling, the
+  /// method will continue sending the update until the scrolling finishes.
```
What is "the update"?

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+  /// If the `continuous` is set to true and the update causes scrolling, the
+  /// method will continue sending the update until the scrolling finishes.
+  ///
+  /// The `continuous` defaults to false if not set.
```
```suggestion
The continuous argument defaults to false.
```

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+    }
+  }
+
+  /// Copy
```
this needs more doc

chunhtai commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,207 @@
+/// the [SelectionHandler] API of the delegate.
+///
+/// To use this widget, set the [registrar] this container received the
+/// [SelectionEvent]s from. Use [SelectionRegistrarScope.maybeOf] to retrieve
```
I am struggling on, too. I know it make it a bit cumbersome use. The reason I do it this way is why would someone want to create a SelectionContainer if there is no selectionRegistrar.

chunhtai committed:
Addressing review comments

chunhtai commented:
@goderbauer @justinmc This pr is ready for another looks. Some highlights from latest commit

1. I moved gestures back to SelectableRegion, and makes all the selection APIs private.
2. Updated SelectionContainer to automatically grab SelectionRegister from context if not provided
3. Not sure whether I should change SelectionRegisterScope.disabled. Should continue discussion on https://github.com/flutter/flutter/pull/95226#discussion_r868476047
4. Not sure if there is action item in https://github.com/flutter/flutter/pull/95226#discussion_r868620538

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+/// ## Selections
+///
+/// To make this [RichText] Selectable, the [RichText] needs to be in the
+/// subtree of a [SelectionArea] or [SelectableRegion] and assigned a
+/// [SelectionRegistrar] to the [RichText.selectionRegistrar]. One can use
```
"and assigned a [SelectionRegistrar] to the" => "and a [SelectionRegsitrar] needs to be assigned to the"

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+/// subtree of a [SelectionArea] or [SelectableRegion] and assigned a
+/// [SelectionRegistrar] to the [RichText.selectionRegistrar]. One can use
+/// [SelectionRegistrarScope.maybeOf] to get the [SelectionRegistrar] from a
+/// context. This enable users to select the texts in [RichText]s with mice or
```
"texts" => "text"

justinmc replied:
"enable" => "enables"

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+/// [SelectionRegistrar] to the [RichText.selectionRegistrar]. One can use
+/// [SelectionRegistrarScope.maybeOf] to get the [SelectionRegistrar] from a
+/// context. This enable users to select the texts in [RichText]s with mice or
+/// touches.
```
Is this better? "touches" => "touch events"

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+///
+/// {@tool snippet}
+///
+/// This sample demonstrates how to assign selection registrar for RichTexts
```
"assign selection registrar" => "assign a [SelectionRegistrar]"

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
````diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+/// ```dart
+/// RichText(
+///   text: const TextSpan(text: 'Hello'),
+///   selectionRegistrar: SelectionRegistrarScope.maybeOf(context),
````
Why is this needed for RichText but not for Text?

chunhtai replied:
Text is smarter than RichText, it will grab the selection registrar/ selection style/ mouse region.  RichText does not automatically imply things.

I can make RichText smarter, but then two thing would happen

1. It can no longer be a RenderObjectWidget because of mouseRegion.
2. customers wouldn't be able to customize these selection behaviors.

justinmc replied:
Got it, I'm on board with it as-is.

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
````diff
@@ -5453,6 +5453,34 @@ class Flow extends MultiChildRenderObjectWidget {
+///   selectionColor: const Color(0xAF6694e8),
+/// )
+/// ```
+/// {@end-tool}
````
Good idea including an example here, this will be helpful for users :+1:

justinmc commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5573,6 +5605,12 @@ class RichText extends MultiChildRenderObjectWidget {
   /// {@macro dart.ui.textHeightBehavior}
   final ui.TextHeightBehavior? textHeightBehavior;
 
+  /// The selection registrar this rich text is subscribed to.
```
"[SelectionRegistrar]"

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -802,6 +820,471 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+  /// [SelectionEdgeUpdateEvent].
+  ///
+  /// When a new [SelectionEdgeUpdateEvent] is dispatched to a selectable, this
+  /// updater checks the current scroll offset against the one stored in this
```
"this" => "these"

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -802,6 +820,471 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+  /// [SelectionEdgeUpdateEvent] and dispatches the event before dispatching the
+  /// new event.
+  ///
+  /// For example, if a selectable receives end [SelectionEdgeUpdateEvent] and
```
"end" => "an end"

justinmc commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -802,6 +820,471 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
+  /// new event.
+  ///
+  /// For example, if a selectable receives end [SelectionEdgeUpdateEvent] and
+  /// its scroll offset in the records is different from the current, it
```
"current" => "current value"

justinmc reviewed (commented):
Mostly nits here.  I plan to do another round of playing around with this on my machine.

chunhtai committed:
update

justinmc commented on packages/flutter/lib/src/material/selection_area.dart:
```diff
@@ -0,0 +1,93 @@
+    TextSelectionControls? controls = widget.selectionControls;
+    switch (Theme.of(context).platform) {
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+        controls ??= materialTextSelectionControls;
```
Ah looks like you're right, definitely no need to worry then.  Elsewhere I've seen Fuchsia considered a desktop platform.

justinmc commented on packages/flutter/test/widgets/selection_container_test.dart:
```diff
@@ -0,0 +1,148 @@
+      expect(delegate.selectables.length, 0);
+    });
+
+    testWidgets('selection container register itself if there is a selectable child', (WidgetTester tester) async {
```
"register" => "registers"

justinmc commented on packages/flutter/test/widgets/selection_container_test.dart:
```diff
@@ -0,0 +1,148 @@
+          ),
+        ),
+      );
+      expect(registrar.selectables.length, 1);
+      expect(delegate.selectables.length, 3);
```
It's not immediately clear to me why these aren't both 3...

chunhtai replied:
`registrar` will be what selection container register to, 
`delegate` will be what the selectable (3 text widgets) in the subtree register too

justinmc replied:
Got it, thanks.

justinmc commented on packages/flutter/test/widgets/selection_container_test.dart:
```diff
@@ -0,0 +1,148 @@
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('SelectionContainer', () {
```
Nit: Why put everything in a group?

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      await gesture.up();
+    }, skip: isBrowser); // https://github.com/flutter/flutter/issues/61020
+
+    testWidgets('touch selection', (WidgetTester tester) async {
```
Nit: Maybe 'long press and drag touch selection'?  It took me a second to realize that's what the test is doing.  Up to you though.

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      final Map<String, dynamic> clipboardData = mockClipboard.clipboardData as Map<String, dynamic>;
+      expect(clipboardData['text'], 'w are you?Fine');
+    },
+      variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.android, TargetPlatform.windows, TargetPlatform.linux, TargetPlatform.fuchsia }),
```
Why does this only apply to these platforms?

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      skip: isBrowser, // https://github.com/flutter/flutter/issues/61020
+    );
+
+    testWidgets('widget span is ignored if it does not contain text', (WidgetTester tester) async {
```
This is a good test, I wasn't really clear about how this situation would work until looking at it.

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      await gesture.up();
+    });
+
+    testWidgets('mouse can copy content', (WidgetTester tester) async {
```
Nit: The copying is actually being done by the keyboard.  Maybe 'can copy a selection made with the mouse'?

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      await gesture.up();
+      expect(
+          renderSelectionSpy.events.every((SelectionEvent element) => element is ClearSelectionEvent),
+          isTrue
```
Comma here, and the indentation is too much I think.

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+    testWidgets('mouse long press does not send select-word event', (WidgetTester tester) async {
+      final UniqueKey spy = UniqueKey();
+      await tester.pumpWidget(
+          MaterialApp(
```
Too much indentation here and above in the pumpWidgets in this file.

justinmc replied:
Actually I'm not even clear if this situation should be 2 or 4 spaces, but either way is fine I guess :man_shrugging:

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      final TestGesture gesture = await tester.startGesture(const Offset(200.0, 200.0));
+      addTearDown(gesture.removePointer);
+      await tester.pump(const Duration(milliseconds: 500));
+      expect(renderSelectionSpy.events.length, 1);
```
It's cool that you can get these events in the test like this :+1:

justinmc commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -0,0 +1,980 @@
+      );
+    });
+
+    testWidgets('mouse selection always cancel previous selection', (WidgetTester tester) async {
```
"cancel" => "cancels"

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,652 @@
+      expect(find.descendant(of: find.text('Item 0'), matching: find.byType(RichText)), findsNothing);
+      expect(find.descendant(of: find.text('Item 1'), matching: find.byType(RichText)), findsNothing);
+
+      // Start copying
```
Period here and on some similar comments above in this file.

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,652 @@
+      innerParagraph24 = tester.renderObject<RenderParagraph>(find.descendant(of: find.text('Iteminner 24'), matching: find.byType(RichText)));
+      expect(innerParagraph23.selections[0], const TextSelection(baseOffset: 2, extentOffset: 12));
+      expect(innerParagraph24.selections[0], const TextSelection(baseOffset: 0, extentOffset: 2));
+    });
```
The nested scrollable thing was tricky, good call testing it.

justinmc commented on packages/flutter/test/widgets/scrollable_selection_test.dart:
```diff
@@ -0,0 +1,652 @@
+
+  setUp(() async {
+    TestDefaultBinaryMessengerBinding.instance!.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.platform, mockClipboard.handleMethodCall);
+    await Clipboard.setData(const ClipboardData(text: 'empty'));
```
Nit: Why the string "empty" if the clipboard is not empty?

chunhtai replied:
I would like to differentiate between clipboard.setData is never called or empty string is written into the clipboard,

justinmc commented on packages/flutter/test/material/selection_area_test.dart:
```diff
@@ -0,0 +1,36 @@
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  testWidgets('Selection area use correct selection controls', (WidgetTester tester) async {
```
'Selection area use' => 'SelectionArea uses'

justinmc commented on packages/flutter/lib/src/widgets/text.dart:
```diff
@@ -342,10 +345,25 @@ class DefaultTextHeightBehavior extends InheritedTheme {
 /// [TapGestureRecognizer] as the [TextSpan.recognizer] of the relevant part of
 /// the text.
 ///
+/// ## Selections
```
Nit: Maybe just "Selection" no plural?

justinmc commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,218 @@
+
+  /// The delegate for [SelectionEvent]s sent to this selection container.
+  ///
+  /// The [Selectable]s in the subtree is added or removed from this delegate
```
"is" => "are"

justinmc commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,218 @@
+  /// The delegate for [SelectionEvent]s sent to this selection container.
+  ///
+  /// The [Selectable]s in the subtree is added or removed from this delegate
+  /// uses [SelectionRegistrar] API.
```
"uses" => "using the"

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,607 @@
+  /// content.
+  ///
+  /// {@macro flutter.rendering.selection.SelectionResult.footNote}
+  // See `_SelectableRegionState._triggerSelectionEndEdgeUpdate` for how this
+  // result affects the selection.
```
Did you mean for this to be triple slash `///`?

chunhtai replied:
This is a comment for code reader, the customer that implements Selectable does not need to know how `pending` is used. They just need to know they should return pending if the selection can't be determine in this frame.

justinmc replied:
Ah I see, sounds good.

justinmc commented on packages/flutter/lib/src/rendering/selection.dart:
```diff
@@ -0,0 +1,607 @@
+}
+
+/// The selected content in a [Selectable] or [SelectionHandler].
+// TODO(chunhtai): Add more support for rich content.
```
Nit: Consider creating and linking issues for this and other TODOs in this PR, if you don't already have a good way to keep track of them.

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1584 @@
+  PointerDeviceKind.invertedStylus,
+};
+
+
```
Extra newline here.

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1584 @@
+};
+
+
+/// A widget that introduces an area for user selections.
```
Nit: Maybe say "selection" without the plural since there can only be one selection at a time?  Up to you though if you think plural is more clear.

chunhtai replied:
although one selection at a time, users can make multiple selections during the life time of this widget, so I think it makes more sense to keep it plural.

justinmc replied:
True, I agree with that :+1:

justinmc reviewed (commented):
I did some more QA and I think I found a few bugs that probably need to be fixed.  Besides that I think the code and docs look good now.  The comments are all either nits or small changes.

### Bug

  1. Run an app with some text and a TextField inside of a SelectionArea, such as the app given below.
  2. Focus the TextField and type some text.
  3. Press ctrl-A to select all.

Expected: Only the text in the TextField is selected.
Actual: The opposite: everything except the text in the TextField is selected.

![Screenshot from 2022-05-17 15-43-56](https://user-images.githubusercontent.com/389558/168924260-25703129-76e9-4156-b074-06f2ca059e29.png)

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: const Text(_title)),
        body: Center(
          child: SelectionArea(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: const <Widget>[
                Text('Selectable text'),
                SelectionRegistrarScope.disabled(child: Text('Non-selectable text')),
                Text('Selectable text'),
                TextField(),
                Text('Selectable text'),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

### Another possible bug

In the same example given above, selecting across the TextField doesn't select the text inside the field.  I would expect it to completely select everything inside the TextField, which is what happens when I try the same thing on a plain HTML page.

![Screenshot from 2022-05-17 15-48-38](https://user-images.githubusercontent.com/389558/168924731-fd59cafe-1f71-42e2-9d7f-ea1acd7dffc0.png)

### Copy didn't work for me sometimes
In the example, if I click and drag to select some of the text but I start the gesture near the edge of the window outside of the Text widgets, then ctrl-c doesn't copy the selected text.  The correct text is selected, and right clicking and using the menu to copy does work, but not ctrl-c.

### Shortcuts/gestures
Maybe some shortcuts and things will come later, but I noticed a few that obviously didn't work:

  * Should the arrow keys work in this PR?  They aren't working for me.
  * Right clicking a word selects the word, but I think it shouldn't.  In a TextField, right clicking just sets the cursor.
  * Double-click to select a word doesn't work.

chunhtai committed:
addressing comments

chunhtai commented:
@justinmc The shortcuts issues are fixed by using the _NonOverrideAction introduced in the last commit. I addressed all of the comments except the following 

>Should the arrow keys work in this PR? They aren't working for me.
No, I plan to add this in the follow up PR.

>Right clicking a word selects the word, but I think it shouldn't. In a TextField, right clicking just sets the cursor.

I couldn't reproduce it. I tested on macOS and it select the whole word in textfield.

>Double-click to select a word doesn't work.
No, I will add that in the follow up. There is already a lot of code in the PR, I think think this may be a good cut-off between milestones.

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1599 @@
+/// An action that does not override any [Action.overridable] in the subtree.
+///
+/// If a intent is fire at or below the context of [Action.overridable] in the
+/// subtree, this action will respect the original action and not override it.
```
What would happen if someone makes _NonOverrideAction overridable? :exploding_head: Is it worth documenting that case?

chunhtai replied:
selectable region already does it. so unfortunately, the user can't tell whether they are override the SelectableRegion or EditableText's actions.

but if they also use _NonOverrideAction, every thing will work as expected. Intent fired from from EditableText's it will fire bubble up to the top most _NonOverrideAction which will call a chain of callingActions back to the EditableText. Intent fired from SelectableRegion, also bubble up to top most _NonOverrideAction and chain of callingAction will only go back to SelectableRegion because SelectableRegion's callingAction will be null

justinmc reviewed (approved):
LGTM :+1:  Let's do it, I'm super excited for this.  Everything below and in my comments are nits or could be done in a follow-up PR.

### A few last things to look at

What do you think about my "Another possible bug" section in my previous review?  It might only be a problem on web, not sure how it should behave on other platforms.  It's something that can be addressed later though.

 >  > Right clicking a word selects the word, but I think it shouldn't. In a TextField, right clicking just sets the cursor.
 >
 > I couldn't reproduce it. I tested on macOS and it select the whole word in textfield.

Oh the reason must be because I'm on Linux.  [Here](https://github.com/flutter/flutter/blob/4c0b0be2da5d1ee80c3d713e68ddd88d2cf2e72d/packages/flutter/lib/src/widgets/text_selection.dart#L1637)'s the code that handles right click in TextSelectionGestureDetectorBuilder, maybe we need to do a similar switch or somehow share the code for global selection.

I'm on board with splitting out the gesture stuff into another PR for sure though, up to you whether you think this right click stuff belongs in this PR or not.

### Everything else is excellent

I tried everything in my QA above and it works great now.  I also tried overriding both global selection and TextField shortcuts with one Shortcuts widget and it worked great (code below).  The same thing but overriding an Intent using an Actions widget also worked great.

<details>

<summary>Example of remapping Shortcut for both global selection and TextField</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Shortcuts(
        shortcuts: const <ShortcutActivator, Intent>{
          SingleActivator(LogicalKeyboardKey.keyB, control: true): SelectAllTextIntent(SelectionChangedCause.keyboard),
        },
        child: SelectionArea(
          child: Scaffold(
            appBar: AppBar(title: const Text(_title)),
            body: Center(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: const <Widget>[
                  Text('Row lorem ipsum 1'),
                  Text('Row lorem ipsum 2'),
                  Text('Row lorem ipsum 3'),
                  TextField(),
                  Text('Row lorem ipsum 4'),
                  Text('Row lorem ipsum 5'),
                  Text('Row lorem ipsum 6'),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

chunhtai commented:
regarding the linux gesture, does it have different behavior when right click on a native web page text vs native web page textfield?

If we want the unify behavior between textfield and global selection, @Renzo-Olivares 's inherited text gesture recognizer can solve this issue. Once it is in, I can also make global selection to grab the text gesture recognizer from the inherited widget.

justinmc commented:
Here's the behavior that I'm seeing on Linux Chrome:

  Right clicking in a <textarea> always selects the word.
  Right clicking on an <input> when focused selects the word.
  Right clicking on an <input> when not focused sets the cursor.
  Right clicking on some non-editable text doesn't select the word.

Putting this off for now sounds good to me.

chunhtai committed:
addressing comments

goderbauer commented on examples/api/lib/material/selection_area/custom_container.dart:
```diff
@@ -0,0 +1,136 @@
+}
+
+class _SelectionAllOrNoneContainerState extends State<SelectionAllOrNoneContainer> {
+  final MySelectAllOrNoneContainerDelegate delegate = MySelectAllOrNoneContainerDelegate();
```
nit: remove the "My" from the name?

goderbauer commented on examples/api/lib/material/selection_area/custom_container.dart:
```diff
@@ -0,0 +1,136 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final SelectionRegistrar? registrar = SelectionRegistrarScope.maybeOf(context);
```
Is this still needed? I thought the SelectionContainer does this automatically now.. could simplify the example a little bit.

goderbauer commented on examples/api/lib/material/selection_area/custom_container.dart:
```diff
@@ -0,0 +1,136 @@
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: _title,
+      home: SelectionArea(
```
I don't think I remember this from my last read-through and before I forget: Do we document somewhere where one should introduce a SelectionArea if one wants to activate global selection? E.g. would it have been ok to just wrap the entire MaterialApp in one? Does this have to be done per route? Or something else?

chunhtai replied:
There are some documentations in both SelectableRegion and SelectionArea to talk about default behavior and how to use them

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+/// {@end-tool}
+///
+/// In the case where a group of widgets should be excluded from selection under
+/// a [SelectableRegion], consider wrapping that group of widgets using
+/// [SelectionRegistrarScope.disabled].
```
static field would mean it could never be const. That would be unfortunate...

goderbauer commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -0,0 +1,1432 @@
+        selection = const TextSelection.collapsed(offset: 1);
+        break;
+    }
+    textEditingValue = TextEditingValue(text: '__', selection: selection);
```
Let's add some TODO comments here explaining that.

chunhtai replied:
There is a TODO at `textEditingValue` property

goderbauer reviewed (commented):
Overall this LGTM after the last few comments are addressed.

chunhtai commented on packages/flutter/lib/src/widgets/selection_container.dart:
```diff
@@ -0,0 +1,218 @@
+        }
+      }
+    }
+    if (widget.registrar != null)
```
they will be handled in didUpdateWidget

chunhtai committed:
Addressing comments

goderbauer reviewed (approved):
LGTM

chunhtai commented on examples/api/lib/material/selection_area/disable_partial_selection.dart:
```diff
@@ -0,0 +1,37 @@
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: const <Widget>[
+                Text('Selectable text'),
+                SelectionRegistrarScope.disabled(child: Text('Non-selectable text')),
```
moved to SelectionContainer.disable

chunhtai committed:
fix build

nilsreichardt commented:
> Right clicking a word selects the word, but I think it shouldn't. In a TextField, right clicking just sets the cursor.

@justinmc On macOS this not true for web (I don't know how Windows it handels):

https://user-images.githubusercontent.com/24459435/170144319-1ee2bad2-50ea-476d-8b1c-6104699d2f29.mov

justinmc commented:
@nilsreichardt Good catch.  I think we're doing those gestures in another PR so we'll have to get this right there.

nilsreichardt commented:
@justinmc Yes, I also opened a few issue for the `SelectionArea` widget which can be implemented in separated PRs 👍  My intention with the comment was the current selection behavior of `SelectionArea` is correct (right click and the word becomes selected) ✅
